lib/ecmascript.ts

```diff
@@ -3233,444 +3375,468 @@ export const ES = ObjectAssign({}, ES2020, {
+  // not expected and so is avoided below via a fast path for time-only
+  // arithmetic.
+  // BTW, this behavior is similar in spirit to offset: 'prefer' in `with`.
+  const TemporalDuration = GetIntrinsic('%Temporal.Duration%');
+  if (DurationSign(years, months, weeks, days, 0, 0, 0, 0, 0, 0) === 0) {
+    return AddInstant(GetSlot(instant, EPOCHNANOSECONDS), h, min, s, ms, µs, ns);
+  }
+
+  // RFC 5545 requires the date portion to be added in calendar days and the
+  // time portion to be added in exact time.
+  const dt = BuiltinTimeZoneGetPlainDateTimeFor(timeZone, instant, calendar);
+  const datePart = CreateTemporalDate(GetSlot(dt, ISO_YEAR), GetSlot(dt, ISO_MONTH), GetSlot(dt, ISO_DAY), calendar);
+  const dateDuration = new TemporalDuration(years, months, weeks, days, 0, 0, 0, 0, 0, 0);
+  const addedDate = CalendarDateAdd(calendar, datePart, dateDuration, options);
+  const dtIntermediate = CreateTemporalDateTime(
+    GetSlot(addedDate, ISO_YEAR),
+    GetSlot(addedDate, ISO_MONTH),
+    GetSlot(addedDate, ISO_DAY),
+    GetSlot(dt, ISO_HOUR),
+    GetSlot(dt, ISO_MINUTE),
+    GetSlot(dt, ISO_SECOND),
+    GetSlot(dt, ISO_MILLISECOND),
+    GetSlot(dt, ISO_MICROSECOND),
+    GetSlot(dt, ISO_NANOSECOND),
+    calendar
+  );
+
+  // Note that 'compatible' is used below because this disambiguation behavior
+  // is required by RFC 5545.
+  const instantIntermediate = BuiltinTimeZoneGetInstantFor(timeZone, dtIntermediate, 'compatible');
+  return AddInstant(GetSlot(instantIntermediate, EPOCHNANOSECONDS), h, min, s, ms, µs, ns);
+}
+
+export function RoundNumberToIncrement(quantity, increment, mode) {
+  if (increment === 1) return quantity;
+  let { quotient, remainder } = quantity.divmod(increment);
+  if (remainder.equals(bigInt.zero)) return quantity;
+  const sign = remainder.lt(bigInt.zero) ? -1 : 1;
+  switch (mode) {
+    case 'ceil':
+      if (sign > 0) quotient = quotient.add(sign);
+      break;
+    case 'floor':
+      if (sign < 0) quotient = quotient.add(sign);
+      break;
+    case 'trunc':
+      // no change needed, because divmod is a truncation
+      break;
+    case 'halfExpand':
+      // "half up away from zero"
+      if (remainder.multiply(2).abs() >= increment) quotient = quotient.add(sign);
+      break;
+  }
+  return quotient.multiply(increment);
+}
+
+export function RoundInstant(epochNs, increment, unit, roundingMode) {
+  // Note: NonNegativeModulo, but with BigInt
+  let remainder = epochNs.mod(86400e9);
+  if (remainder.lesser(0)) remainder = remainder.plus(86400e9);
+  const wholeDays = epochNs.minus(remainder);
+  const roundedRemainder = RoundNumberToIncrement(remainder, nsPerTimeUnit[unit] * increment, roundingMode);
+  return wholeDays.plus(roundedRemainder);
+}
+
+export function RoundISODateTime(
+  year,
+  month,
+  day,
+  hour,
+  minute,
+  second,
+  millisecond,
+  microsecond,
+  nanosecond,
+  increment,
+  unit,
+  roundingMode,
+  dayLengthNs: bigInt.BigInteger | number = 86400e9
```

Comment:
Is it valid to assign a non-`n` literal to a bigint?  Something seems fishy here. Seems like it should either be a number or a bigint, but mixing them seems to invite bugs.  Not saying this needs to be fixed in this PR-- probably better somewhere else--but at least wanted to flag it here.

Reply:
This pre-existing fishiness is my fault for being in a hurry to get ZonedDateTime working... I never had time to track down all the places where this was inconsistent. Using tsc it should be easier to fix, hopefully.

Reply:
> Is it ever possible for the length of a day to be longer than number-size nanoseconds?

No. MAX_SAFE_INTEGER nanoseconds is 104 days. 

I'm not sure about using bigint internally. Others may have opinions about this. 

I assume a reasonable rule should be this: if the value is only used in calculations with bigints, then make it a bigint like you did with DAYNANOS. Otherwise, then it's probably a case-by-case decision.